File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Ruby/climbing_stairs.rb
# You are climbing a stair case. It takes n steps to reach to the top.
# Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

# @param {Integer} n
# # @return {Integer}
def climb_stairs(n)
  # For each step n, the number of ways is the sum of ways of (n - 1) and (n - 2)
  step_ways = [0, 1, 2]
  iter = 3
  while iter <= n
    step_ways[iter] = step_ways[iter - 1] + step_ways[iter - 2]
    iter += 1
  end
  return step_ways[n]
end

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/ClimbingStairs.cpp
/*

You are climbing a stair case. It takes n steps to reach to the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

*/

class Solution {
public:
    int climbStairs(int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if( n < 1 )
            return -1;
            
        vector<int> steps;
        steps.push_back(1);
        steps.push_back(2);
        
        if( n < 3 )
            return steps[n-1];
            
        int temp_step;
        for( int i = 2; i < n; i++){
            temp_step = steps[i-1] + steps[i-2];
            steps.push_back( temp_step );
        }
        
        return steps[n-1];   
    }
};

void main(){

  Solution test;
  test.climbStairs(5);
}
